src/main/java/org/elasticsearch/index/mapper/internal/IndexFieldMapper.java

```diff
@@ -157,6 +170,62 @@ public String value(Object value) {
     }
 
     @Override
+    public Filter termsFilter(List values, @Nullable QueryParseContext context) {
+        if (fieldType.indexed() || context == null){
+            super.termsFilter(values, context);
+        }
+        for (Object value : values) {
+            if (isSameIndex(value, context)) {
+                return Queries.MATCH_ALL_FILTER;
+            }
+        }
+        return Queries.MATCH_NO_FILTER;
+    }
+
+
+    /**
+     * This termQuery impl looks at the context to determine the index that is being queried and
+     * then creates returns a MATCH_ALL_FILTER or MATCH_NO_FILTER if the value matches this index
+     * this can be useful if aliases or wildcards are used but the aim is to restrict the query to specific indices
+     */
+    @Override
+    public Query termQuery(Object value, @Nullable QueryParseContext context) {
+        if (fieldType.indexed() || context == null) {
+            return super.termQuery(value, context);
+        }
+        if (isSameIndex(value, context)) {
+            return new XConstantScoreQuery(Queries.MATCH_ALL_FILTER);
+        } else {
+            return new XConstantScoreQuery(Queries.MATCH_NO_FILTER);
+        }
```

Comment:
I can't see a check anywhere in here for the special case where the field name is "_index". Without this surely there is the potential for non-indexed fields such as "foo" to do a MATCH_ALL if the value of foo happens to match the index name?